src/gluonnlp/model/bert.py

```diff
@@ -407,22 +412,35 @@ def _apply_pooling(self, sequence):
         outputs = sequence[:, 0, :]
         return self.pooler(outputs)
 
-    def _decode(self, sequence, valid_length=None):
+    def _decode(self, sequence, positions):
         """Generate unnormalized prediction for the masked language model task.
 
         This is only used for pre-training the BERT model.
+
+        Inputs:
+            - **sequence**: input tensor of sequence encodings.
+              Shape (batch_size, seq_length, units).
+            - **position**: input tensor of position of tokens for decoding.
+              Shape (batch_size, max_num_positions)
+
+        Outputs:
+            - **decode_outputs**: output tensor of token predictions for target positions.
+                Shape (batch_size, max_num_positions, vocab_size).
         """
-        if valid_length is None:
-            last_step = sequence[:, -1, :]
-        else:
-            batch_size = sequence.shape[0]
-            ctx = valid_length.context
-            dtype = valid_length.dtype
-            batch_idx = mx.nd.arange(0, batch_size, dtype=dtype, ctx=ctx).reshape((1, -1))
-            valid_length = (valid_length - 1).reshape((1, -1))
-            last_step_idx = mx.nd.Concat(batch_idx, valid_length, dim=0)
-            last_step = mx.nd.gather_nd(sequence, last_step_idx)
-        return self.decoder(last_step)
+        batch_size = sequence.shape[0]
+        max_num_positions = positions.shape[1]
```

Comment:
Does the orginal implementation also force `max_num_positions' to be the same in each batch? This seems to be not correct.